graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+	// as source
+	if value.IsValid() && !value.Type().Comparable() {
+		// TODO: Warn, or somehow prevent using type-system?
+		key.source = new(byte)
+	}
+	return key
+}
+
+// resolveNonBatchUnit resolves a non-batch field in our graphql response graph.
+func resolveNonBatchUnit(ctx context.Context, src interface{}, dest *objectWriter, unit *ExecutionUnit) []*ExecutionUnit {
+	fieldResult, err := unit.field.Resolve(ctx, src, unit.selection.Args, unit.selection.SelectionSet)
+	if err != nil {
+		dest.Fail(err)
+		return nil
+	}
+	subFieldExecUnits, err := unwrapBatchResult(ctx, []interface{}{fieldResult}, unit.field.Type, unit.selection.SelectionSet, []*objectWriter{dest})
+	if err != nil {
+		dest.Fail(err)
+		return nil
+	}
+	return subFieldExecUnits
+}
+
+// unwrapBatchResult traverses the provided sources and fills in result data and
+// returns new execution units that are required to resolve the rest of the
+// query result.
+func unwrapBatchResult(ctx context.Context, sources []interface{}, typ Type, selectionSet *SelectionSet, destinations []*objectWriter) ([]*ExecutionUnit, error) {
+	switch typ := typ.(type) {
+	case *Scalar:
+		return nil, unwrapBatchScalarResult(sources, typ, destinations)
+	case *Enum:
+		return nil, unwrapBatchEnumResult(sources, typ, destinations)
+	case *List:
+		return unwrapBatchListResult(ctx, sources, typ, selectionSet, destinations)
+	case *Union:
+		return unwrapBatchUnionResult(ctx, sources, typ, selectionSet, destinations)
+	case *Object:
+		return unwrapBatchObjectResult(ctx, sources, typ, selectionSet, destinations)
+	case *NonNull:
+		return unwrapBatchResult(ctx, sources, typ.Type, selectionSet, destinations)
+	default:
+		panic(typ)
+	}
+}
+
+// Resolves the scalar type value for all the provided sources.
+func unwrapBatchScalarResult(sources []interface{}, typ *Scalar, destinations []*objectWriter) error {
+	for i, source := range sources {
+		if typ.Unwrapper == nil {
+			destinations[i].Fill(unwrap(source))
+			continue
+		}
+		res, err := typ.Unwrapper(source)
+		if err != nil {
+			return err
+		}
+		destinations[i].Fill(res)
+	}
+	return nil
+}
+
+// Resolves the enum type value for all the provided sources.
+func unwrapBatchEnumResult(sources []interface{}, typ *Enum, destinations []*objectWriter) error {
+	for i, source := range sources {
+		val := unwrap(source)
+		if mapVal, ok := typ.ReverseMap[val]; !ok {
+			err := errors.New("enum is not valid")
+			destinations[i].Fail(err)
+			return err
+		} else {
+			destinations[i].Fill(mapVal)
+		}
+	}
+	return nil
+}
+
+// Flattens the sources for the list type and calls into an unwrapper method for
+// the list's subtype.
+func unwrapBatchListResult(ctx context.Context, sources []interface{}, typ *List, selectionSet *SelectionSet, destinations []*objectWriter) ([]*ExecutionUnit, error) {
+	flattenedResps := make([]*objectWriter, 0, len(sources))
+	flattenedSources := make([]interface{}, 0, len(sources))
```

Comment:
allocate for sum len of sources?

Reply:
Added an iteration step to grab the proper length first.